optaplanner-test/src/main/java/org/optaplanner/test/impl/score/stream/DefaultConstraintVerifier.java

```diff
@@ -1,63 +1,79 @@
 
-    private final ServiceLoader<ScoreDirectorFactoryService> serviceLoader;
     private final ConstraintProvider_ constraintProvider;
     private final SolutionDescriptor<Solution_> solutionDescriptor;
-    private ConstraintStreamImplType constraintStreamImplType;
-    private Boolean droolsAlphaNetworkCompilationEnabled;
+    /**
+     * {@link ConstraintVerifier} is mutable,
+     * due to {@link #withConstraintStreamImplType(ConstraintStreamImplType)} and
+     * {@link #withDroolsAlphaNetworkCompilationEnabled(boolean)}.
+     * Since these methods can be run at any time, possibly invalidating the pre-built score director factories,
+     * the easiest way of dealing with the issue is to keep an internal immutable constraint verifier instance
+     * and clearing it every time the configuration changes.
+     * The code that was using the old configuration will continue running on the old instance,
+     * which will eventually be garbage-collected.
+     * Any new code will get a new instance with the new configuration applied.
+     */
+    private final AtomicReference<ConfiguredConstraintVerifier<ConstraintProvider_, Solution_, Score_>> configuredConstraintVerifierRef =
+            new AtomicReference<>();
+    private final AtomicReference<ConstraintStreamImplType> constraintStreamImplTypeRef = new AtomicReference<>();
+    private final AtomicReference<Boolean> droolsAlphaNetworkCompilationEnabledRef = new AtomicReference<>();
 
     public DefaultConstraintVerifier(ConstraintProvider_ constraintProvider, SolutionDescriptor<Solution_> solutionDescriptor) {
-        this.serviceLoader = ServiceLoader.load(ScoreDirectorFactoryService.class);
         this.constraintProvider = constraintProvider;
         this.solutionDescriptor = solutionDescriptor;
     }
 
     public ConstraintStreamImplType getConstraintStreamImplType() {
-        return constraintStreamImplType;
+        return constraintStreamImplTypeRef.get();
     }
 
     @Override
     public ConstraintVerifier<ConstraintProvider_, Solution_> withConstraintStreamImplType(
             ConstraintStreamImplType constraintStreamImplType) {
         requireNonNull(constraintStreamImplType);
-        this.constraintStreamImplType = constraintStreamImplType;
+        var droolsAlphaNetworkCompilationEnabled = this.droolsAlphaNetworkCompilationEnabledRef.get();
+        if (droolsAlphaNetworkCompilationEnabled != null &&
+                droolsAlphaNetworkCompilationEnabled &&
+                constraintStreamImplType != ConstraintStreamImplType.DROOLS) {
+            throw new IllegalArgumentException("Can not switch to " + ConstraintStreamImplType.class.getSimpleName()
+                    + "." + constraintStreamImplType + " while Drools Alpha Network Compilation enabled.");
+        }
+        this.constraintStreamImplTypeRef.set(constraintStreamImplType);
+        this.configuredConstraintVerifierRef.set(null);
```

Comment:
A thread B might still get the old `ConfiguredConstraintVerifier` instance if it enters the `verifyThat()` while thread A is between these two lines.
I don't think it's a problem, as even with full synchronization, the result just depends on which method gets called earlier (`withConstraintStreamImplType` or `verifyThat`).